### ISSUE TITLE
error 042: invalid block header when performing backup

### ISSUE THREAD

**Issue by fishermd13** [pgbackrest/pgbackrest#403]
We have several large postgres clusters that are backed up nightly (incremental with periodic full backups).  Occasionally, a backup will fail with a variation of the following output:

<pre>
ERROR [042]: : raised on 'local-2' host: invalid block header '{"err":42,"out":"unable to deflate '/var/lib/postgresql/9.5/main/base/26564/2167677.1': stream error"}'
ERROR [042]: : process 'local-2' terminated unexpectedly [000]:         (in cleanup) malformed JSON string, neither array, object, number, string or atom, at character offset 0 (before "BRBLOCK0") at /usr/share/perl5/pgBackRest/Protocol/Base/Master.pm line 165.
 at /usr/bin/pgbackrest line 15.
        main::__ANON__("malformed JSON string, neither array, object, number, string "...) called at /usr/share/perl/5.22/Carp.pm line 166
        Carp::croak("malformed JSON string, neither array, object, number, string "...) called at /usr/share/perl/5.22/JSON/PP.pm line 1209
        JSON::PP::decode_error("malformed JSON string, neither array, object, number, string "...) called at /usr/share/perl/5.22/JSON/PP.pm line 1041
        JSON::PP::word() called at /usr/share/perl/5.22/JSON/PP.pm line 729
        JSON::PP::value() called at /usr/share/perl/5.22/JSON/PP.pm line 689
        JSON::PP::PP_decode_json(JSON::PP=HASH(0x26b1b50), "BRBLOCK0", 0) called at /usr/share/perl/5.22/JSON/PP.pm line 148
        JSON::PP::decode(JSON::PP=HASH(0x26b1b50), "BRBLOCK0") called at /usr/share/perl5/pgBackRest/Protocol/Base/Master.pm line 165
        pgBackRest::Protocol::Base::Master::outputRead(undef) called at /usr/share/perl5/pgBackRest/Protocol/Storage/File.pm line 135
        pgBackRest::Protocol::Storage::File::close(pgBackRest::Protocol::Storage::File=HASH(0x2ab83c8)) called at /usr/share/perl5/pgBackRest/Common/Io/Base.pm line 118
        pgBackRest::Common::Io::Base::DESTROY(pgBackRest::Protocol::Storage::File=HASH(0x2ab83c8)) called at /usr/share/perl/5.22/Carp.pm line 167
        eval {...} called at /usr/share/perl/5.22/Carp.pm line 167
        pgBackRest::Backup::File::backupFile() called at /usr/share/perl5/pgBackRest/Protocol/Local/Minion.pm line 57
        pgBackRest::Protocol::Local::Minion::__ANON__(ARRAY(0x2ab7ab0)) called at /usr/share/perl5/pgBackRest/Protocol/Base/Minion.pm line 146
        eval {...} called at /usr/share/perl5/pgBackRest/Protocol/Base/Minion.pm line 163
        eval {...} called at /usr/share/perl5/pgBackRest/Protocol/Base/Minion.pm line 171
        pgBackRest::Protocol::Base::Minion::process(pgBackRest::Protocol::Local::Minion=HASH(0x22ac138)) called at /usr/bin/pgbackrest line 130
        eval {...} called at /usr/bin/pgbackrest line 275

        (in cleanup) pgBackRest::Common::Exception=HASH(0x3a61958) at /usr/share/perl/5.22/Carp.pm line 167.
</pre>

or

<pre>
ERROR [042]: : raised on 'local-3' host: invalid block header '{"out":"unable to deflate '/var/lib/postgresql/9.5/main/base/26564/29020.38': stream error","err":42}'
ERROR [042]: : process 'local-3' terminated unexpectedly [000]: 	(in cleanup) malformed JSON string, neither array, object, number, string or atom, at character offset 0 (before "BRBLOCK0") at /usr/share/perl5/pgBackRest/Protocol/Base/Master.pm line 165.
 at /usr/bin/pgbackrest line 15.
	main::__ANON__("malformed JSON string, neither array, object, number, string "...) called at /usr/share/perl/5.22/Carp.pm line 166
	Carp::croak("malformed JSON string, neither array, object, number, string "...) called at /usr/share/perl/5.22/JSON/PP.pm line 1209
	JSON::PP::decode_error("malformed JSON string, neither array, object, number, string "...) called at /usr/share/perl/5.22/JSON/PP.pm line 1041
	JSON::PP::word() called at /usr/share/perl/5.22/JSON/PP.pm line 729
	JSON::PP::value() called at /usr/share/perl/5.22/JSON/PP.pm line 689
	JSON::PP::PP_decode_json(JSON::PP=HASH(0x2e51f78), "BRBLOCK0", 0) called at /usr/share/perl/5.22/JSON/PP.pm line 148
	JSON::PP::decode(JSON::PP=HASH(0x2e51f78), "BRBLOCK0") called at /usr/share/perl5/pgBackRest/Protocol/Base/Master.pm line 165
	pgBackRest::Protocol::Base::Master::outputRead(undef) called at /usr/share/perl5/pgBackRest/Protocol/Storage/File.pm line 135
	pgBackRest::Protocol::Storage::File::close(pgBackRest::Protocol::Storage::File=HASH(0x3659848)) called at /usr/share/perl5/pgBackRest/Common/Io/Base.pm line 118
	pgBackRest::Common::Io::Base::DESTROY(pgBackRest::Protocol::Storage::File=HASH(0x3659848)) called at /usr/share/perl/5.22/Carp.pm line 167
	eval {...} called at /usr/share/perl/5.22/Carp.pm line 167
	pgBackRest::Backup::File::backupFile() called at /usr/share/perl5/pgBackRest/Protocol/Local/Minion.pm line 57
	pgBackRest::Protocol::Local::Minion::__ANON__(ARRAY(0x3659800)) called at /usr/share/perl5/pgBackRest/Protocol/Base/Minion.pm line 146
	eval {...} called at /usr/share/perl5/pgBackRest/Protocol/Base/Minion.pm line 163
	eval {...} called at /usr/share/perl5/pgBackRest/Protocol/Base/Minion.pm line 171
	pgBackRest::Protocol::Base::Minion::process(pgBackRest::Protocol::Local::Minion=HASH(0x2a4b0f8)) called at /usr/bin/pgbackrest line 130
	eval {...} called at /usr/bin/pgbackrest line 275

	(in cleanup) pgBackRest::Common::Exception=HASH(0x2bb9f70) at /usr/share/perl/5.22/Carp.pm line 167.
2017-08-24 10:13:40.410 P00   INFO: backup command end: aborted with exception [042]
</pre>

This isn't limited to just one cluster and has occurred on PostgreSQL 9.3 and 9.5 databases.
The command line used to initiate the backup is:
(although "--stop-auto" was a recent addition and this has occurred with and without it)

`/usr/bin/pgbackrest --stop-auto --stanza=ourstanzasname --type=incr backup`

The filename referenced varies, but always seems to be the "last" file in the directory...  The one that grows in size during operation.

Manually running the same (incr) command later will sometimes be successful, otherwise I'll run a full backup for it to complete without error.
I don't see anything noteworthy in the postgres logs which would indicate any problems or file corruption, and I've run a filesystem check on the target to make sure there was no physical damage to the device.

/etc/pgbackrest.conf
<pre>
[ourstanzasname]
db1-host=hostname1
db1-path=/var/lib/postgresql/9.5/main
db1-user=postgres
db2-host=hostname2
db2-path=/var/lib/postgresql/9.5/main
db2-user=postgres

[global]
backup-standby=y
repo-path=/var/lib/pgbackrest
retention-full=2
cmd-ssh=/usr/bin/ssh
process-max=4
</pre>

Thanks!

**Comment by dwsteele** [comment]
It looks like the core error is:
```
unable to deflate '/var/lib/postgresql/9.5/main/base/26564/29020.38': stream error
```
Everything else is just errors that get thrown because the protocol is in a bad state at shutdown.

I'm not sure what is causing the gzip compression to fail.  It's likely bad inputs but I haven't figured out how that's happening yet.  This appears to be the same issue as #404 so I'm guessing it was introduced recently.

What version of pgBackRest are you running?  Did you upgrade from a previous version where things were working? If so, what version what that?

**Comment by fishermd13** [comment]
My apologies for the version omission.  We're currently running 1.20.  I did not remember seeing this problem in 1.18.

On Monday I'll upgrade to 1.22 to see if the problem exists there (unless the reporter of issue 404 follows up, saying they are running 1.22)

**Comment by dwsteele** [comment]
Thanks for the information - it's what I suspected.  The relevant code was heavily refactored in 1.19 when S3 support was added and I presume that's when the issue started.

I don't think upgrading to 1.22 is likely to help.  You might consider reverting to 1.18 instead while we tackle the problem.  It sounds like this is causing you some operational headaches...

**Comment by fishermd13** [comment]
I appreciate your time and the response.  I've rolled back to 1.18 and will follow up if the problem turns out to go back that far.

I seem to remember having to change the configuration setting for archive-queue-max after one of the upgrades.  Do you happen to know if "archive-queue-max=1073741824" will be valid for the older version?  I can't seem to track down an earlier version of the docs.

Thanks again!

**Comment by dwsteele** [comment]
`archive-queue-max` was introduced in 1.13 so you should be good.  Let me know how it goes.

**Comment by dwsteele** [comment]
Also, what OS/Version are you on?  I'm trying to determine if this is a zlib issue - or at least a version compatibility issue.

**Comment by fishermd13** [comment]
The dedicated backup server is running Ubuntu 16.04 and the database boxes are running Ubuntu 14.04 and 16.04 (all of which are 64-bit).  Both are using zlib1g version 1.2.8.

**Comment by mydimension** [comment]
I'm getting similar errors when performing restores (non-production machine thankfully):

```2017-08-27 14:11:33.627 P00  ERROR: [042]: raised on 'local-1' host: invalid block header '{"out":[{"pgBackRest::Common::Io::Handle":1034207232}]}'
2017-08-27 14:11:33.629 P00  ERROR: [042]: raised on 'local-1' host: invalid block header '{"out":[{"pgBackRest::Common::Io::Handle":1034207232}]}'
2017-08-27 14:42:07.239 P00  ERROR: [042]: process 'local-8' terminated unexpectedly [000]:     (in cleanup) pgBackRest::Common::Exception=HASH(0x39f40d0) at /usr/share/perl/5.22/Carp.pm line 167.

2017-08-27 14:42:19.260 P00  ERROR: [042]: process 'local-7' terminated unexpectedly [000]:     (in cleanup) pgBackRest::Common::Exception=HASH(0x44ce998) at /usr/share/perl/5.22/Carp.pm line 167.

2017-08-27 14:42:54.090 P00  ERROR: [042]: process 'local-5' terminated unexpectedly [000]:     (in cleanup) pgBackRest::Common::Exception=HASH(0x35a03a8) at /usr/share/perl/5.22/Carp.pm line 167.

2017-08-27 14:42:54.107 P00  ERROR: [042]: process 'local-4' terminated unexpectedly [000]:     (in cleanup) pgBackRest::Common::Exception=HASH(0x489c6f0) at /usr/share/perl/5.22/Carp.pm line 167.

2017-08-27 14:42:54.109 P00  ERROR: [042]: process 'local-3' terminated unexpectedly [042]: unable to read line after 1830 second(s) from stdio
                                        process 'pgbackrest-server remote' terminated unexpectedly [042]: unable to read line after 1830 second(s) from stdio


2017-08-27 14:42:54.110 P00   WARN: unable to shutdown protocol [042]: process 'local-3' terminated unexpectedly [042]: unable to read line after 1830 second(s) from stdio
                                        process 'pgbackrest-server remote' terminated unexpectedly [042]: unable to read line after 1830 second(s) from stdio


2017-08-27 14:42:54.126 P00  ERROR: [042]: process 'local-2' terminated unexpectedly [000]:     (in cleanup) pgBackRest::Common::Exception=HASH(0x36edad8) at /usr/share/perl/5.22/Carp.pm line 167.

2017-08-27 14:42:54.143 P00  ERROR: [042]: process 'local-1' terminated unexpectedly [000]:     (in cleanup) pgBackRest::Common::Exception=HASH(0x317b8a0) at /usr/share/perl/5.22/Carp.pm line 167.

2017-08-27 14:42:54.200 P00  ERROR: [042]: process 'pgbackrest-server remote' terminated unexpectedly [042]: unable to read line after 1830 second(s) from stdio

2017-08-27 14:42:54.201 P00   WARN: unable to shutdown protocol [042]: process 'pgbackrest-server remote' terminated unexpectedly [042]: unable to read line after 1830 second(s) from stdio

2017-08-27 14:42:54.202 P00  ERROR: [042]: raised on 'local-1' host: invalid block header '{"out":[{"pgBackRest::Common::Io::Handle":1034207232}]}'
2017-08-27 14:42:54.203 P00   INFO: restore command end: aborted with exception [042]
```

command is:

```
pgbackrest restore --stanza main --delta
```

spartan local config that merely points at the backup server.

backup server config:

```
[global]
repo-path=/var/lib/pgbackrest
log-level-file=detail
process-max=8
hardlink=y
start-fast=y
stop-auto=y
compress=n
retention-full=2
retention-diff=1

[main]
backup-standby=y
db1-path=/var/lib/postgresql/9.4/main
db1-host=primary-server
db2-path=/var/lib/postgresql/9.4/main
db2-host=replica-server
```

**Comment by dwsteele** [comment]
@mydimension This is certainly a regression introduced in 1.19 but is not related to this issue.  The keep-alives are not being propagated to the remotes properly in the new scheme so the remotes time out while the delta process is ongoing.  We'll fix this in the next release, but in the meantime you can set `protocol-timeout` to a value longer than you expect the restore to take.

**Comment by fishermd13** [comment]
Granted, it's only been a few days, but every backup since reverting back to 1.18 has been successful.  I'll follow up again if I see the same error--but it looks like you were right about it starting in 1.19.

**Comment by dwsteele** [comment]
Thanks for the info.  I have been unsuccessful reproducing this error no matter how I poke at the compression routines.  I'm beginning to suspect that it may be an issue in the Perl-C interface but have not been able to device a test to demonstrate it yet.  I'll keep you posted.

**Comment by dwsteele** [comment]
@mydimension I have created #408 to track the issue you found.

**Comment by dwsteele** [comment]
@fishermd13 OK, looks like I have found a fix for this.  The compression loop was continuing to read even after a zero read flushed the compression buffer.  If there *was* more to read then the compression state was bad and a stream error was thrown on deflate.

How would you like your name to appear in the release notes?

**Comment by fishermd13** [comment]
@dwsteele Excellent!  Thank you for the work and the amazing product.

No specific mention is necessary, sir, but I appreciate the ask.

**Comment by dwsteele** [comment]
Let me know if you change your mind -- I like to give credit where it is due.  Your analysis was a great help in tracking down the problem.

**Comment by dwsteele** [comment]
Fixed at 206415d4c75.
